### PR TITLE
574 - Support for Dark mode

### LINKED ISSUE #574: [Feature] Support a Dark Mode theme
### Description of Feature

Add support for dark mode.

- [ ] Allow the user to choose the theme (find somewhere in the UI to put a toggle)
- [ ] Detect the user's preference

### PR DESCRIPTION
**Pull request by tinahollygb**
### Features and Changes

Adds support for dark mode that:

- uses the users default operating system setting (if available)
- allows the user to override their operating system setting

I've gone through all of the screens that I could find. For ones I couldn't find, I made a best guess based on the colours I saw. 

Changes:

- Bootstrap components, as well as CSS-in-JS components needed to be overridden. 

Things I didn't change:

- most things with purple. So buttons remain as they were. Also the sidebar remained unchanged as it is dark mode-friendly enough.

Issues:

- Closes #574 

### How to be Dark-mode friendly as a developer

I will write a proper document on Notion but for now, the TL;DR is:

1. use the CSS variables, i.e. `var(--text-color-main)` instead of `$black` so that it will flip properly
2. when adding a new library that uses CSS, test it in dark mode and ensure it looks good. if it doesn't have the right theme, override the colours as I did (see the `bootstrap-theme-overrides` file).
3. if you need to add a new colour, add it to both the light and dark mixins in the `theme.scss` file.
4. test and when putting up a PR, include screenshots of both light and dark modes

### Dependencies

n/a

### Testing

Test all the things.

#### Latest Icons

Latest icons (not reflected in below screenshots).

image
image
image

#### Top Bar

CleanShot 2022-09-30 at 12 38 43@2x

image

image

#### Sidebar

No changes to the theme, just the dropdowns to be dark-friendly.

image
image

#### Forms and Alerts

Forms were styled to be dark-friendly.

CleanShot 2022-09-30 at 12 45 14@2x

Alerts too. 
Screen Shot 2022-09-29 at 6 30 18 PM
Screen Shot 2022-09-29 at 6 32 13 PM
Screen Shot 2022-09-29 at 6 36 50 PM

#### Features

image
image

#### Experiments

image

We can maybe improve the placeholder images by using transparent images.

image

image

Diff viewer.

image
image

#### Metrics

image
image

#### Segments

image

#### Dimensions

image

#### Data Sources

image
image
image

For the Code component that uses AceEditor, I opted for theme `tomorrow_night` instead of `textmate` for dark mode.

image

image

image

#### Import config.yml differ

image

#### Management

image

image
image
image
image

#### Settings

image
image

#### Visualizations

image
image

image
image
image
image

image
image

Draw
image

Warning
image

#### Getting Started flow

Expand the toggle section below to see all the screenshots.

<details>
<summary><b>Getting Started section</b></summary>

image
image
image
image
image
image

image
image
image

image

</details>

### CODE REVIEW

File: packages/front-end/components/Forms/CodeTextArea.tsx [hunk 1]
```diff
@@ -49,6 +53,28 @@ export default function CodeTextArea({
...
+
+      if (fromStorage === "dark") {
+        actualTheme = DARK_THEME;
+      }
+    } catch (e) {
+      return actualTheme;
+    }
+
+    return actualTheme;
+  }, []);
```
Comment (tinahollygb) [inline packages/front-end/components/Forms/CodeTextArea.tsx:76]: When we need to do something like this again, it would be a good time to do a refactor to add it as a context provider and expose a `useAppearanceTheme()` hook or similar.

File: packages/front-end/components/Layout/ThemeToggler.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      if (theme === "system") {
+        localStorage.removeItem("gb_ui_theme");
+      } else {
+        document.documentElement.classList.add(`theme--${theme}`);
+        localStorage.setItem("gb_ui_theme", theme);
+      }
+    } catch (e) {
+      // We are unable to persist the theme changes due to the browser's privacy settings
+    }
+  }, []);
```
Comment (tinahollygb) [inline packages/front-end/components/Layout/ThemeToggler.tsx:63]: Related to the last comment, we would refactor this into a context provider and this component would just be UI that links up the functions from that provider.

File: packages/front-end/styles/_bootstrap-theme-overrides.scss [hunk 3]
```diff
@@ -0,0 +1,372 @@
+// Global Bootstrap overrides that respect dark/light modes
```
Comment (tinahollygb) [inline packages/front-end/styles/_bootstrap-theme-overrides.scss:1]: These are mostly Bootstrap but we have other components that are overridden here.

File: packages/front-end/styles/_bootstrap-theme-overrides.scss [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+  }
+
+  [class$="-multiValue"] {
+    background-color: var(--form-multivalue-background-color);
+
+    > * {
+      color: var(--form-multivalue-text-color);
+    }
+  }
+}
```
Comment (tinahollygb) [inline packages/front-end/styles/_bootstrap-theme-overrides.scss:82]: These selectors are quite fragile. If we update the library we use for multiple select, there's a good chance this will break. That said, if we do that, it's likely the developer would test that and we'd catch it, so probably ok to leave them.

File: packages/front-end/styles/_bootstrap-theme-overrides.scss [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+  [class*="-gutter"][class*="-added"] {
+    background-color: var(--diff-added-background-highlight);
+    color: var(--text-color-main);
+  }
+}
+
+// region d2h differ
+
+// d2h loads CSS at the end of the document so we need to do all these !important
```
Comment (tinahollygb) [inline packages/front-end/styles/_bootstrap-theme-overrides.scss:321]: This is the differ used in the import config.yml section.

File: packages/front-end/styles/_bootstrap-theme-overrides.scss [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+.nav-tabs .nav-link:focus {
+  border-color: var(--border-color-300) var(--border-color-300)
+    var(--surface-background-color);
+}
+
+// endregion Tabs
+
+// region Diff Viewer
+
+.diff-wrapper {
```
Comment (tinahollygb) [inline packages/front-end/styles/_bootstrap-theme-overrides.scss:266]: This is the differ used in the history panel.

File: packages/front-end/styles/_bootstrap-variable-overrides.scss [hunk 7]
```diff
@@ -0,0 +1,13 @@
+$font-size-base: 0.875rem; // change the default from 16px to 14px.
```
Comment (tinahollygb) [inline packages/front-end/styles/_bootstrap-variable-overrides.scss:1]: Original overrides moved to a separate file.

File: packages/front-end/components/Layout/ThemeToggler.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+  return (
+    <div className="dropdown top-nav-theme-menu">
+      <button
+        onClick={() => {
+          setThemeDropdownOpen(!themeDropdownOpen);
+        }}
+        className="btn mr-1 text-secondary"
+      >
+        <FaMoon />
```
Comment (jdorn) [inline packages/front-end/components/Layout/ThemeToggler.tsx:73]: Should we change the icon here depending on what theme they have selected?  Also, I think we should try to give some indication that this is clickable and what it does.  Could be as simple as adding a dropdown arrow.  Or a title/tooltip saying "Change theme".

File: packages/front-end/components/Forms/CodeTextArea.tsx (hunk 1 above)
Comment (jdorn) [inline packages/front-end/components/Forms/CodeTextArea.tsx:76]: We need this same logic in the `Code` component.  It already supports light/dark mode, but the theme is passed in as a prop right now instead of using the actual selected theme.

It's currently using the "light" theme to render JSON on the feature page and the SQL/Python code for data sources:

image
image


And it's using the "dark" theme when rendering code samples and SQL queries on the metric page:

image

image

**Review by jdorn — COMMENTED** [review]
Found a few edge cases in dark mode:

### Upgrade modal on Cloud
Screenshot from 2022-10-04 07-35-17

To test this, you'll need to render it manually somewhere:

```tsx
      <UpgradeModal
        close={() => {console.log("close")}}
        source="top-nav-freeseat-overage"
        reason="Whoops! You are over your free seat limit."
      />
```

### Activity Feed
Click the Bell icon in the top nav
Screenshot from 2022-10-04 07-31-46

### Review Feature Changes
When you make changes to a feature and review them before publishing
Screenshot from 2022-10-04 07-30-11

### Namespace Settings
First, add a namespace in Settings.  Then, add an experiment rule to a feature and enable namespaces.  The legend and purple number are really hard to read in dark mode.
image

**Comment by tinahollygb** [comment]
Updated the loading spinner in the modal.

image

I've also updated the dropdown icon to reflect the currently-selected theme:

image
image
image

To support the multiple areas where we need to get the current theme, I refactored the theme logic into a context provider.

File: packages/front-end/components/Forms/CodeTextArea.tsx (hunk 1 above)
Comment (tinahollygb) [reply packages/front-end/components/Forms/CodeTextArea.tsx]: I've updated all code blocks to use the theme from the context provider.

CleanShot 2022-10-04 at 14 08 53@2x
CleanShot 2022-10-04 at 14 09 50@2x

image
image

File: packages/front-end/components/Layout/ThemeToggler.tsx (hunk 8 above)
Comment (tinahollygb) [reply packages/front-end/components/Layout/ThemeToggler.tsx]: I've updated the toggler to change the icon based on the preferred theme: system, dark, light.

File: packages/front-end/styles/theme.scss [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+// OS default
+:root:not(.theme--light):not(.theme--dark) {
+  @media (prefers-color-scheme: light) {
+    @include light-theme();
+  }
+
+  @media (prefers-color-scheme: dark) {
+    @include dark-theme();
+  }
```
[the remark this answers is not in the document]
Comment (tinahollygb) [reply packages/front-end/styles/theme.scss:125]: `prefers-color-scheme` has been [available since 2019](https://caniuse.com/prefers-color-scheme) in Chrome, Edge, Firefox, Safari and Opera so I think it should be ok, but I've defaulted to light just in case.

**Comment by tinahollygb** [comment]
I've updated the code fold section:

image

The review feature changes diff has been updated. It uses the `ExpandableDiff` component. So far I've found 3 different components for diffs. The project implements 2 diff plugins so far. Do you know if we have any more instances of diffs than the ones I've already found? If they implement the same one this section uses, it'll need to be wrapped in `diff-wrapper` to have the changes applied.

image

**Comment by tinahollygb** [comment]
I have updated the namespace background too.

image

**Comment by tinahollygb** [comment]
I've also improved the placeholder image appearance.

image
image

This was the original, for reference. The following old images are now outdated:

CleanShot 2022-10-04 at 14 41 38@2x
CleanShot 2022-10-04 at 14 42 06@2x

While the contrast is poor in dark mode, I think it's preferred over the bright image, so I've made the change. Let me know if you'd like me to revert back to the previous placeholder style.

**Comment by tinahollygb** [comment]
@jdorn I believe I have implemented the change requests. You can see individual comments with screenshots but here is a summary:

- fixed the dark mode modal loading state overlay
- fixed the code folding area in the activity feed diff
- added the `diff-wrapper` class to the `ExpandableDiff` component so it could use the defined styles 
- improved background and border colours for the namespace visualization
- defaulted to light theme should `prefers-color-scheme` not be available
- implemented the dynamic theme in the `<Code />` block, which updates all instances to use either light or dark depending on the theme chosen. This will have the change that in light mode it will use the light theme, which will be different than previous behaviour.
- refactored the theme logic into a context provider to expose a hook so we can get the current theme (light, dark) as well as the user-preferred theme (system, light, dark).
- adjusted the placeholder image background—this causes it to have low contrast in dark mode but it's less jarring

Let me know if I missed anything or if there's anything else that needs to be changed.

**Review by jdorn — DISMISSED** [review]

**Review by jdorn — APPROVED** [review]

**Review by Auz — APPROVED** [review]
